File: python/src/providers/openai/generate.py
import httpx

from common.config import Settings, require_openai


def generate(prompt: str, settings: Settings) -> str:
    require_openai(settings)
    url = "https://api.openai.com/v1/chat/completions"
    headers = {
        "Authorization": f"Bearer {settings.openai_api_key}",
        "Content-Type": "application/json",
    }
    payload = {
        "model": settings.openai_model,
        "messages": [{"role": "user", "content": prompt}],
        "temperature": 0.2,
    }

    response = httpx.post(url, headers=headers, json=payload, timeout=settings.timeout_s)
    response.raise_for_status()
    data = response.json()
    try:
        content = data["choices"][0]["message"]["content"]
    except (KeyError, IndexError, TypeError) as exc:
        raise RuntimeError("Unexpected response from OpenAI.") from exc
    return content.strip()

File: node/src/providers/openai/generate.mjs
import { requireOpenAI } from "../../common/config.js";

const CHAT_COMPLETIONS_URL = "https://api.openai.com/v1/chat/completions";

/**
 * generate sends a single-turn prompt to OpenAI chat completions.
 */
export async function generate(prompt, settings) {
  requireOpenAI(settings);
  const payload = {
    model: settings.openaiModel,
    messages: [{ role: "user", content: prompt }],
    temperature: 0.2,
  };

  const data = await postJson(CHAT_COMPLETIONS_URL, payload, settings);
  const content = data?.choices?.[0]?.message?.content;
  if (!content) {
    throw new Error("Unexpected response from OpenAI.");
  }
  return content.trim();
}

async function postJson(url, payload, settings) {
  const controller = new AbortController();
  const timeoutId = setTimeout(() => controller.abort(), settings.timeoutMs);

  try {
    const response = await fetch(url, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${settings.openaiApiKey}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(payload),
      signal: controller.signal,
    });

    const text = await response.text();
    if (!response.ok) {
      throw new Error(`OpenAI request failed (${response.status}): ${text}`);
    }

    return JSON.parse(text);
  } finally {
    clearTimeout(timeoutId);
  }
}